File: cpp/hot100/32.随机链表的复制.cpp
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head==nullptr) return nullptr;
        Map<Node,Node> map=new HashMap<>();
        Node curr=head;
        while(curr){
            map.put(curr,new Node(curr.val));
            curr=curr->next;
        }
        curr=head;
        while(curr){
            Node clone=map.get(curr);
            clone->next=map.get(curr->next);
            clone->random=map.get(curr->random);
            curr=curr->next;
        }
        return map.get(head);
    }
};
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head==nullptr){
            return nullptr;
        }
        for(Node* cur=head;cur;cur=cur->next->next){
            cur->next=new Node(cur->val,cur->next,nullptr);
        }
        for(Node *cur=head;cur;cur=cur->next->next){
            if(cur->random) cur->next->random=cur->random->next;
        }
        Node *new_head=head->next;
        Node *cur=head;
        for(;cur->next->next;cur=cur->next){
            Node *copy=cur->next;
            cur->next=copy->next;
            copy->next=copy->next->next;
        }
        cur->next=nullptr;
        return new_head;
    }
};

File: python/hot100/32.随机链表的复制.py
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        nodedict={}
        cur=head
        while cur:
            nodedict[cur]=Node(cur.val)
            cur=cur.next
        cur=head
        while cur:
            clone=nodedict[cur]
            clone.next=nodedict.get(cur.next)
            clone.random=nodedict.get(cur.random)
            cur=cur.next
        return nodedict[head]
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return None
        cur=head
        while cur:
            cur.next=Node(cur.val,cur.next)
            cur=cur.next.next
        cur=head
        while cur:
            if cur.random:
                cur.next.random=cur.random.next
            cur=cur.next.next
        cur=head.next
        while cur.next:
            cur.next=cur.next.next
            cur=cur.next
        return head.next